### PR TITLE
docs(security): extension permissions docs, manifest invariants, messaging trust boundary

### LINKED ISSUE #676: Chrome extension permissions の利用理由を docs 化し、権限追加時の確認テストを追加する
## 優先度

P1

## 背景

TABBIN の manifest では `alarms` / `tabs` / `storage` / `contextMenus` / `notifications` と、local Ollama 用の host permissions を利用している。

現状の機能上、多くは必要な権限だが、Chrome拡張では permissions が増えるほどストア審査・ユーザー説明・セキュリティレビューの重要度が上がる。

各 permission の利用理由を docs 化し、将来 permission が増えたときにレビューしやすくしたい。

## 対象ファイル

- `wxt.config.ts`
- generated Chrome / Firefox manifest
- `docs/security/permissions.md` など新規docs
- background messaging / runtime protocol
- 必要に応じて manifest validation test

## 現在の権限候補

- `alarms`
- `tabs`
- `storage`
- `contextMenus`
- `notifications`
- `host_permissions: http://localhost:11434/*`
- `host_permissions: http://127.0.0.1:11434/*`

## 対応方針

- 各 permission の利用機能・利用ファイル・必要理由を docs にまとめる
- local Ollama への host permissions は外部送信ではなく local runtime 接続であることを明記する
- permission 追加時に docs 更新を促せるよう、manifest permission snapshot test を検討する

## generated manifest の security invariant を検証する

WXT config の source 定義だけでなく、最終的な Chrome / Firefox generated manifest を対象に validation test を追加する。

検証候補:

- `permissions` allowlist
- `host_permissions` allowlist
- `externally_connectable` が意図せず追加されていない
- `content_scripts` が意図せず追加されていない
- `web_accessible_resources` allowlist
- CSP invariant
- `unsafe-eval` 不在
- remote script を前提とする設定不在
- Chrome / Firefox manifest delta が想定範囲内

新しい privileged surface を追加した場合、snapshot 更新だけで通さず security review が必要なことを docs に明記する。

## runtime messaging の trust boundary を定義する

現在の internal runtime messaging は Zod schema で payload validation しているが、将来 `content_scripts` / `externally_connectable` / `onMessageExternal` を追加すると sender trust が変わる。

以下を明文化する。

- extension page / background 間の internal message
- content script 由来 message
- external web page / extension 由来 message
- privileged action の sender policy

以下の追加を manifest / architecture test で検知し、message trust boundary review を要求する。

```text
content_scripts 追加
externally_connectable 追加
onMessageExternal 利用追加
  ↓
trust boundary review 必須
```

privileged action候補:

- URL / storage record 削除
- timestamp更新
- auto-delete実行
- AI request実行
- 将来のimport / repair / migration command

現在直ちにすべてのinternal messageでsender allowlistを追加することを前提にせず、sender typeごとのtrust levelと必要なvalidationを定義してから実装する。

## docs イメージ

```md
# Chrome extension permissions

## tabs

### 用途
- 現在タブ・ウィンドウ内タブ・同一ドメインタブの取得
- 保存済みタブを開く

### 利用箇所
- `src/lib/background/extension-actions.ts`
- `src/lib/background/url-storage.ts`

## storage

### 用途
- savedTabs / urls / userSettings / customProjects の保存
```

## 受け入れ条件

- [ ] 権限ごとの利用理由が docs 化されている
- [ ] 各 permission の主な利用ファイルが記載されている
- [ ] local Ollama host permission の目的が明記されている
- [ ] permission 追加時に docs 更新を促す仕組みがある
- [ ] generated Chrome / Firefox manifest を validation する
- [ ] `permissions` / `host_permissions` を allowlist 検証する
- [ ] `externally_connectable` / `content_scripts` / `web_accessible_resources` の security invariant を検証する
- [ ] CSP / remote script に関する invariant を検証する
- [ ] Chrome / Firefox manifest delta を確認できる
- [ ] runtime messaging の sender type ごとの trust boundary が docs 化されている
- [ ] privileged action の sender policy を定義している
- [ ] `content_scripts` / `externally_connectable` / `onMessageExternal` 追加時に trust boundary review を促す仕組みがある
- [ ] 既存テストが通る

## 注意点

- 権限を削ること自体が目的ではなく、利用理由と security boundary の透明化が目的
- 実際に不要な権限が見つかった場合は別Issueで削除する
- Firefox向け manifest 差分も検証対象にする
- manifest snapshot の機械的更新だけで新しい privileged surface を承認しない
- internal message と untrusted content-script / external message を同じ trust level として扱わない

### PR DESCRIPTION
**Pull request by haruto-yamada1**
## Cause

Issue #676 requires documenting why each Chrome extension permission is requested, adding generated-manifest security invariant tests, and defining the runtime messaging trust boundary. The generated-manifest verifier previously checked the permission/host-permission allowlist and CSP, but did not guard `externally_connectable`, `content_scripts`, `web_accessible_resources`, or the Chrome/Firefox manifest delta. There were no docs for per-permission usage or the messaging sender trust model.

## Changes

- `docs/security/permissions.md` (new): documents each approved API permission (`alarms`, `tabs`, `storage`, `contextMenus`, `notifications`, `unlimitedStorage`) and host permission (`http://localhost:11434/*`, `http://127.0.0.1:11434/*`) with purpose, primary usage sites, and review notes; documents the generated-manifest security invariants and the checklist for adding a new permission or privileged surface.
- `docs/security/messaging-trust-boundary.md` (new): defines sender types and trust levels (extension page/background trusted; content script and external senders untrusted), privileged action sender policy, trust-boundary review triggers, and the migration control key trust boundary framework (references #727 as the authoritative state-machine owner).
- `tools/scripts/manifestSecurityInvariants.ts` (new): extracted manifest invariant logic. Adds `assertGeneratedManifestSecurityInvariants` (rejects `externally_connectable`, non-empty `content_scripts`, and non-empty/unapproved `web_accessible_resources`) and `assertChromeFirefoxManifestDelta` (fails divergence outside the expected manifest-version-driven keys). CSP verification (`assertExtensionCspMatchesProductionNetworkPolicy`) moved here.
- `tools/scripts/manifestHelpers.ts` (new): shared `isRecord` / `isHostPermission` / `readStringArray` helpers, breaking the import cycle between the two modules.
- `tools/scripts/production-network-policy.ts`: wires the new invariant checks into `assertManifestMatchesProductionNetworkPolicy`; re-exports nothing (consumers import the delta verifier directly). Under the `max-lines` limit after extraction.
- `tools/scripts/verify-production-network-policy.ts`: runs the Chrome/Firefox delta check after per-manifest verification.
- `src/lib/architecture/messagingTrustBoundary.test.ts` (new): source-level guard that detects `onMessageExternal` / `onConnectExternal` listeners in production source and `externally_connectable` / `content_scripts` declarations in `wxt.config.ts`, and asserts the two security docs and the verifier cover the required surfaces.
- `tools/scripts/production-network-policy.test.ts`: adds tests for the new invariants (externally_connectable, content_scripts, web_accessible_resources string/object/malformed forms, absent WAR, WAR allowlist empty) and the delta verifier (valid divergence, API/host divergence, unexpected key, MV3/MV2 version checks, non-object manifests).

## Acceptance criteria mapping

- [x] 権限ごとの利用理由が docs 化されている — `docs/security/permissions.md`
- [x] 各 permission の主な利用ファイルが記載されている — per-permission "Primary usage sites"
- [x] local Ollama host permission の目的が明記されている — host permissions section + `production-network-policy.md` link
- [x] permission 追加時に docs 更新を促す仕組みがある — verifier fails loudly + "Adding a new permission" checklist + `APPROVED_WEB_ACCESSIBLE_RESOURCES` allowlist
- [x] generated Chrome / Firefox manifest を validation する — `assertManifestMatchesProductionNetworkPolicy` + `assertChromeFirefoxManifestDelta`
- [x] `permissions` / `host_permissions` を allowlist 検証 — existing + retained
- [x] `externally_connectable` / `content_scripts` / `web_accessible_resources` の security invariant を検証 — new invariants
- [x] CSP / remote script に関する invariant を検証 — CSP directive pinning (script-src, connect-src, object-src, frame-src, form-action; no unexpected directives)
- [x] Chrome / Firefox manifest delta を確認できる — `assertChromeFirefoxManifestDelta`
- [x] runtime messaging の sender type ごとの trust boundary が docs 化されている — `docs/security/messaging-trust-boundary.md`
- [x] privileged action の sender policy を定義している — privileged actions table + sender policy
- [x] `content_scripts` / `externally_connectable` / `onMessageExternal` 追加時に trust boundary review を促す仕組みがある — manifest verifier + source-level architecture guard
- [x] 既存テストが通る — `bun run test` green

### Comment-derived criteria (dependencies)

- `unlimitedStorage` allowlist: #735 is CLOSED and adopted `unlimitedStorage`; it is in the approved allowlist and documented in `permissions.md` + `persistence-durability.md`.
- Migration control key trust boundary: #727 is still OPEN. This PR documents the security-invariant and review-trigger side and references #727 as the authoritative state-machine/repository-boundary owner. The source-level guard tracks the framework; the concrete storage access boundary lands with #727.

## Verification

- `bun run test` — all tests pass (node + dom projects)
- `bun run test:coverage` — new modules (`manifestSecurityInvariants.ts`, `manifestHelpers.ts`) at 100% statement/line/function coverage; repo-wide coverage unchanged/improved
- `bun run compile` — clean
- `bun run lint` — clean (no import cycle, under `max-lines`)
- `bun run quality:check` — green (format, lint, compile, test, knip, duplication, secretlint, arch:check)
- `bun run build && bun run build:firefox && bun run verify:production-network-policy` — both generated manifests pass per-manifest invariants and the Chrome/Firefox delta check
- `bun run release:check` — green

## Risk

Docs- and test-only change; no runtime behavior change. The new manifest invariants are enforced against generated artifacts, so a future change that adds `content_scripts`, `externally_connectable`, or `web_accessible_resources` will fail verification until a trust-boundary review updates the allowlist and docs — this is the intended gate.

Closes #676

## Summary by CodeRabbit

- **ドキュメント**
  - メッセージングの信頼境界、権限、特権アクションの確認方針を整理しました。
  - 権限追加時のレビュー手順と、Chrome/Firefox向けマニフェストの監査項目を追加しました。

- **セキュリティ強化**
  - 生成マニフェストの権限、CSP、外部接続、公開リソースを厳格に検証します。
  - Chrome版とFirefox版の許可設定に不整合がないか確認します。

- **テスト**
  - 信頼境界およびマニフェスト検証の網羅的なテストを追加しました。

### CODE REVIEW

**Comment by haruto-yamada1** [comment]
@coderabbit review

File: tools/scripts/manifestSecurityInvariants.ts [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+  // way, an empty array or absence exposes nothing. A non-empty array must be
+  // on the approved allowlist (currently empty) and requires a security review.
+  if ('web_accessible_resources' in manifest) {
+    const resources = manifest.web_accessible_resources
+    if (!Array.isArray(resources) || resources.length !== 0) {
+      throw new Error(
+        `${label} web_accessible_resources must be empty and match the approved allowlist (see docs/security/permissions.md); found ${JSON.stringify(resources)}`,
+      )
+    }
+  }
```
[the remark this answers is not in the document]
Comment (haruto-yamada1) [reply tools/scripts/manifestSecurityInvariants.ts:308]: 指摘を採用しました (commit 16608d88)。

`web_accessible_resources` 検証を `APPROVED_WEB_ACCESSIBLE_RESOURCES` allowlist へ接続しました。MV2 文字列配列と MV3 `{ resources, matches }` オブジェクト配列をリソースパスへ正規化し、allowlist と一致比較します。allowlist を更新すれば承認済みリソースが通り、不一致・超過・不正形式は拒否されます。`docs/security/permissions.md` の `APPROVED_WEB_ACCESSIBLE_RESOURCES` 参照先も `manifestSecurityInvariants.ts` へ修正し、`matches` / `extension_ids` の exposure scope は security review 対象と明記しました。

検証: `bun run quality:check` / `bun run build && bun run build:firefox && bun run verify:production-network-policy` / `bun run release:check` すべて green。新規 test は `assertWebAccessibleResourcesOnAllowlist` で non-empty allowlist 許可・不一致拒否・MV3 object form・不正 MV3 エントリ をカバーします。

File: tools/scripts/manifestSecurityInvariants.ts [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+  for (const key of sharedKeys) {
+    if (expectedDivergentKeys.has(key)) {
+      continue
+    }
+    if (sortedJson(chromeManifest[key]) !== sortedJson(firefoxManifest[key])) {
+      throw new Error(
+        `Chrome and Firefox manifest diverge on unexpected key "${key}": chrome=${sortedJson(chromeManifest[key])} firefox=${sortedJson(firefoxManifest[key])}`,
+      )
+    }
+  }
```
[the remark this answers is not in the document]
Comment (haruto-yamada1) [reply tools/scripts/manifestSecurityInvariants.ts:473]: 指摘を採用しました (commit 16608d88)。

`assertChromeFirefoxManifestDelta` で `action` / `browser_action`, `background`, `browser_specific_settings` を丸ごと skip せず正規化して比較します。
- `action` / `browser_action`: 許可 field (`default_title` / `default_icon` / `default_popup`) のみ許可し想定外 field を拒否。両 browser の正規化 surface を比較し、cross-version key の誤用 (Chrome が `browser_action` 宣言等) を検出。
- `background`: `service_worker` → `scripts` へ正規化して全体構造を比較し、`type` / `persistent` 等の browser 固有追加 field の diverge を検出。
- `browser_specific_settings`: Firefox-only とし Chrome の宣言を拒否、Firefox は `EXPECTED_FIREFOX_BROWSER_SPECIFIC_SETTINGS` と一致比較して構造変更を検出。

検証: 上記と同じ gate すべて green。test は action surface diverge・cross-version key・background extra field・`service_worker` / `scripts` 型 check・`browser_specific_settings` 構造変更・Chrome 宣言拒否 を追加しました。
